lib/chef/provider/aws_cloudsearch_domain.rb

```diff
@@ -0,0 +1,126 @@
+require 'chef/provisioning/aws_driver/aws_provider'
+
+class Chef::Provider::AwsCloudsearchDomain < Chef::Provisioning::AWSDriver::AWSProvider
+
+  def create_aws_object
+    converge_by "Creating new CloudSearch domain #{new_resource.domain_name}" do
+      create_domain
+    end
+    # multi_az defaults to false on creation
+    # so we only need to hit the API if it is true
+    if new_resource.multi_az
+      converge_by "Setting availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Setting scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Setting access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      converge_by "Defining index fields for CloudSearch domain #{new_resource.domain_name}" do
+        new_resource.index_fields.each do |idx_field|
+          if !idx_field.is_a? Hash
+            raise ArgumentError, "Expected index_fields to contain only Hashes"
+          else
+            create_index_field(idx_field)
+          end
+        end
+      end
+    end
+  end
+
+  def destroy_aws_object(instance)
+    converge_by "Deleting CloudSearch domain #{new_resource.domain_name}" do
+      cs_client.delete_domain(domain_name: new_resource.domain_name)
+    end
+  end
+
+  def update_aws_object(instance)
+    if update_availability_options?
+      converge_by "Updating availability options for CloudSearch domain #{new_resource}" do
+        update_availability_options
+      end
+    end
+
+    if update_scaling_params?
+      converge_by "Updating scaling parameters for CloudSearch domain #{new_resource.domain_name}" do
+        update_scaling_parameters
+      end
+    end
+
+    if update_policy?
+      converge_by "Updating access policy for CloudSearch domain #{new_resource.domain_name}" do
+        update_service_access_policy
+      end
+    end
+
+    if update_index_fields?
+      Chef::Log.warn("Updating existing index_fields not currently supported")
+    end
+  end
+
+
+  # TODO(ssd): Expand these to be idempotent?  It takes an extra API
+  # call just to see the current settings so it feels cheaper to just
+  # set them anyway, but then you get updated resources.
+  def update_availability_options?
+    ! new_resource.multi_az.nil?
+  end
+
+  def update_scaling_params?
+    ! (new_resource.partition_count.nil? || new_resource.replication_count.nil? || new_resource.instance_type.nil?)
+  end
+
+  def update_policy?
+    ! new_resource.access_policies.nil?
+  end
+
+  def update_index_fields?
+    ! new_resource.index_fields.nil?
+  end
+
+  def create_domain
+    cs_client.create_domain(domain_name: new_resource.domain_name)
```

Comment:
Per http://docs.aws.amazon.com/AWSRubySDK/latest/AWS/CloudSearch/Client.html#create_domain-instance_method I think this will need to inspect the `:domain_status` field to get the returned object.  That should always be `in_process` or whatever the actual status is.